ovirt_imageio/client/_upload.py

```diff
@@ -0,0 +1,156 @@
+             "block based storage domain (iSCSI, FC). ISO images are "
+             "always uploaded to preallocated disk.")
+
+    cmd.add_argument(
+        "--id", "--disk-id",
+        type=_options.UUID,
+        help="A UUID for the new disk. If not specified oVirt will "
+             "create a new UUID.")
+
+    cmd.add_argument(
+        "filename",
+        type=_options.File,
+        help="Path to image to upload. Supported formats: raw, qcow2, iso.")
+
+
+def upload_disk(args):
+    with _ui.ProgressBar(phase="inspecting image") as progress:
+        disk_info = _probe_image(args)
+        con = _ovirt.connect(args)
+        with closing(con):
+            progress.phase = "creating disk"
+            disk = _ovirt.add_disk(
+                con=con,
+                name=disk_info.name,
+                initial_size=disk_info.initial_size,
+                provisioned_size=disk_info.provisioned_size,
+                sd_name=args.storage_domain,
+                id=args.id,
+                sparse=disk_info.sparse,
+                enable_backup=disk_info.format == _ovirt.COW,
+                content_type=disk_info.content_type,
+                format=disk_info.format)
+
+            progress.phase = "creating transfer"
+            host = _ovirt.find_host(con, args.storage_domain)
+            transfer = _ovirt.create_transfer(con, disk, host=host)
+            try:
+                progress.phase = "uploading image"
+                _api.upload(
+                    args.filename,
+                    transfer.transfer_url,
+                    args.cafile,
+                    buffer_size=args.buffer_size,
+                    progress=progress,
+                    proxy_url=transfer.proxy_url,
+                    max_workers=args.max_workers)
+            except Exception:
+                progress.phase = "cancelling transfer"
+                _ovirt.cancel_transfer(con, transfer)
+                raise
+
+            progress.phase = "finalizing transfer"
+            _ovirt.finalize_transfer(con, transfer, disk)
+        progress.phase = "upload completed"
+
+
+def _probe_image(args):
+    # Obtain the image info dictionary.
+    img_info = _api.info(args.filename)
+    if img_info["format"] not in ("qcow2", "raw"):
+        raise RuntimeError(f"Unsupported image format {img_info['format']}")
+    # Obtain all remaining fields for DiskInfo.
+    initial_size = None
+    if args.format == "qcow2":
+        initial_size = _api.measure(args.filename, args.format)["required"]
+    content_type = _detect_disk_content_type(
+        img_info["filename"], img_info["format"])
+    name = os.path.splitext(os.path.basename(img_info["filename"]))[0]
+    if content_type == _ovirt.ISO:
+        # ISO images require raw format to work with cdrom device
+        format = _ovirt.RAW
+        sparse = False
+    else:
+        format = _ovirt.RAW if args.format == "raw" else _ovirt.COW
+        sparse = args.sparse
```

Comment:
blank line

Reply:
Ended up removing this condition block and splitting it into another helper function:
```python
def _select_disk_format(content_type, disk_format):
```
Hopefully didn't make the code worse with this change, but I thought it may be nice to have it separated.
This is a similar case as with the name, could be obtained from outside this function, but is nice to have them all in one place.